ReSharper.FSharp/src/FSharp.TypeProviders.Host/src/Cache/SharedProvidedCache.cs

```diff
@@ -1,29 +1,37 @@
-﻿using System.Collections.Generic;
+﻿using System.Collections;
+using System.Collections.Generic;
 using System.Linq;
 #if NETFRAMEWORK
 using JetBrains.Collections;
 #endif
 using JetBrains.ReSharper.Plugins.FSharp.TypeProviders.Protocol.Models;
 using JetBrains.ReSharper.Plugins.FSharp.TypeProviders.Protocol.Utils;
+using JetBrains.Util;
+using JetBrains.Util.Collections;
+using JetBrains.Util.dataStructures;
 using static FSharp.Compiler.ExtensionTyping;
 
 namespace JetBrains.ReSharper.Plugins.FSharp.TypeProviders.Host.Cache
 {
-  public abstract class SharedProvidedCache<T> : IBiDirectionalProvidedCache<T, int>
+  public class SharedProvidedCache<T> :
```

Comment:
Could you add a comment why some types can be shared and other can't? Would be even better if the cache name told something about it, e.g. something like:
```
/// Cache holding types coming from referenced assemblies metadata
/// and not depending on particular type provider instances.
///
/// Sharing provided types is not allowed, since type provider instantiations
/// depend on project configuration and may produce different types.
public class MetadataTypesCache
```

It's not exactly clear what does "shared provided" mean for a cache name.